### ISSUE TITLE
Separate Display and Input packages for XCB platforms.

### ISSUE THREAD

**Issue by huba** [kurocha/display-xcb#1]
**Problem:** In the X window system protocol every event is relative to a window.

**Possible Solution:** It is possible to create what's called an `InputOnly` window, which is a window that can never be displayed, but still can get input events. The only way to get input events to it is to grab the pointer or the keyboard it seems. Having to remap mouse pointer events to your actual window seems like a rather messy path to go down, whereas the X server already maps mouse pointer locations to the window they went to (as well as giving you the location relative to root).

The `InputOnly` window will also need to share the same connection to the X server as the `InputOutput` windows (normal displayed ones). Since it's the connection to the server that identifies an application, and it seems you can't just grab input meant for other applications. (for good reason probably)

In conclusion, input and window allocations go hand in hand in X window system clients, one is pretty much useless without the other which is why separate Input and Display focused packages don't make sense for XCB.

**Comment by ioquatix** [comment]
This is all really good.

So the input package hasn't got any platform specific parts. It's just a general set of data structures for handling input, which you expect each display specific adaptor to use.

I expect we should rename this package to `display-xcb` and it should follow the same model as `display-cocoa`.

The key points should be:

- `Display::XCB::Application` should contain the `_connection`.
- `Display::XCB::Window` should contain the `_window`.
- Calling `application.run` should wait for events and dequeue them. Not sure if we can make this async and also cross platform, so it might need to be a blocking operation on the main thread.

Regarding per-window events, I think this makes sense and fits into how `display-cocoa` works. Events are not per-application but per-window and relative to the bounds of the view.

One thing which is probably tricky is mouse grabbing - i.e. hiding the cursor and warping the mouse to the centre of the window, then disassociating movement from cursor location. This is important for "FPS" style games.

Finally, handling full screen is also something worth considering as it is highly WM specific.

**Comment by ioquatix** [comment]
The events should go back to the Window instance which implements `Input::Handler`.
